### PR TITLE
sync start command with cosmos-sdk 0.43

### LINKED ISSUE #340: server start command out of sync with cosmos-sdk 0.43
__Proposal:__ update server start command with cosmos-sdk 0.43 one

__Current behavior:__ out of sync with 0.43

__Desired behavior:__ in sync with 0.43

__Use case:__
- new flags
- flags config file
- grpc web server
- tx/tm services https://github.com/tharsis/ethermint/issues/172

### PR DESCRIPTION
**Pull request by yihuang**
Closes #340

- add grpc web and rosetta server
- add tx/tm services
- add standalone mode
- update cosmos-sdk to 0.43.0-rc2, which fixed a bug in service startup
- add EnableUnsafeCORS option to evm rpc server

## Description

______

For contributor use:

- [x] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [x] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added  relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the Github PR explorer

______

For admin use:

- [ ] Added appropriate labels to PR (ex. `WIP`, `R4R`, `docs`, etc)
- [ ] Reviewers assigned
- [ ] Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

### CODE REVIEW

**Review by fedekunze — COMMENTED** [review]
ACK thanks! minor comment

File: server/start.go [hunk 1]
```diff
@@ -165,10 +190,70 @@ which accepts a path for the resulting pprof file.
...
+
+// legacyAminoCdc is used for the legacy REST API
 func startInProcess(ctx *server.Context, clientCtx client.Context, appCreator types.AppCreator) error {
 	cfg := ctx.Config
 	home := cfg.RootDir
 	logger := ctx.Logger
+	var cpuProfileCleanup func()
+
+	if cpuProfile := ctx.Viper.GetString(flagCPUProfile); cpuProfile != "" {
+		f, err := os.Create(cpuProfile)
```
[the remark this answers is not in the document]
Comment (yihuang) [reply server/start.go:241]: `ExpandHome` seems make sense here, maybe we can push it to upstream.

File: cmd/ethermintd/root.go [hunk 2]
```diff
@@ -52,7 +53,8 @@ func NewRootCmd() (*cobra.Command, params.EncodingConfig) {
 		WithAccountRetriever(types.AccountRetriever{}).
 		WithBroadcastMode(flags.BroadcastBlock).
 		WithHomeDir(app.DefaultNodeHome).
-		WithViper("") // In simapp, we don't use any prefix for env variables.
+		WithKeyringOptions(hd.EthSecp256k1Option()).
+		WithViper("ETHERMINT")
```
[the remark this answers is not in the document]
Comment (yihuang) [reply cmd/ethermintd/root.go]: Yeah, prefix of env variable, like `ETHERMINT_*`.

File: server/start.go [hunk 3]
```diff
@@ -212,153 +304,114 @@ func startInProcess(ctx *server.Context, clientCtx client.Context, appCreator ty
...
-
-		httpSrv = &http.Server{
-			Addr:    config.EVMRPC.RPCAddress,
-			Handler: handlerWithCors.Handler(r),
-		}
+		clientCtx := clientCtx.
+			WithHomeDir(home).
+			WithChainID(genDoc.ChainID)
 
+		apiSrv = api.New(clientCtx, ctx.Logger.With("module", "api-server"))
```
[the remark this answers is not in the document]
Comment (yihuang) [reply server/start.go]: done

File: server/start.go [hunk 4]
```diff
@@ -165,10 +190,70 @@ which accepts a path for the resulting pprof file.
...
+	}
+
+	app := appCreator(ctx.Logger, db, traceWriter, ctx.Viper)
+
+	svr, err := abciserver.NewServer(addr, transport, app)
+	if err != nil {
+		return fmt.Errorf("error creating listener: %v", err)
+	}
+
+	svr.SetLogger(ctx.Logger.With("module", "abci-server"))
```
[the remark this answers is not in the document]
Comment (yihuang) [reply server/start.go]: done

File: cmd/ethermintd/root.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (yihuang) [reply cmd/ethermintd/root.go]: done

File: cmd/ethermintd/config/config.go [hunk 5]
```diff
@@ -74,6 +74,8 @@ type EVMRPCConfig struct {
 	RPCAddress string `mapstructure:"address"`
 	// Address defines the WebSocket server to listen on
 	WsAddress string `mapstructure:"ws-address"`
+	// EnableUnsafeCORS defines if CORS should be enabled (unsafe - use it at your own risk)
+	EnableUnsafeCORS bool `mapstructure:"enable-unsafe-cors"`
```
[the remark this answers is not in the document]
Comment (yihuang) [reply cmd/ethermintd/config/config.go:78]: fixed

File: server/start.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (yihuang) [reply server/start.go]: https://github.com/cosmos/cosmos-sdk/pull/9755

**Review by fedekunze — APPROVED** [review]
ACK

**Comment by fedekunze** [comment]
seems that the solidity tests are not running for some reason

**Comment by yihuang** [comment]
> seems that the solidity tests are not running for some reason

fixed, there was issue with chain id in clientCtx.

**Pull request merged**